src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,187 @@
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getPriceByAddress(address: string, currency: string): Promise<TokenPrice>;
+  getPricesByAddress(addresses: string[], currency: string): Promise<TokenPrice[]>;
+}
+
+// It's convenient to map TokenPrice objects by their address, but consumers typically want an array
+// of TokenPrice objects, so the address must also be embedded within the TokenPrice object. To
+// avoid storing multiple copies of the same TokenPrice, always use the lower-case variant of the
+// address when performing a lookup into the PriceCache, and substitute the provided address into the
+// TokenPrice object when returning to the caller.
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  protected prices: {
+    [currency: string]: PriceCache;
+  } = {};
+
+  constructor(protected logger: Logger, readonly priceFeeds: PriceFeedAdapter[]) {
+    assert(logger, "No logging instance supplied.");
+    assert(priceFeeds.length > 0, "No price feeds supplied.");
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache: PriceCache = this.getPriceCache(currency);
+    const now: number = msToS(Date.now());
+
+    let tokenPrice: TokenPrice | undefined = priceCache[address.toLowerCase()];
+    const cacheMiss = tokenPrice === undefined || now - this.maxPriceAge > tokenPrice.timestamp;
+
+    if (this.maxPriceAge > 0) {
+      const age: number = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_SAFE_INTEGER;
+      this.logger.debug({
+        at: "PriceClient#getPriceByAddress",
+        message: `Cache ${cacheMiss ? "miss" : "hit"} (${currency}) for token ${address}.`,
+        age: `${age} S`,
+        price: tokenPrice,
+      });
+    }
+
+    if (cacheMiss) {
+      const prices: TokenPrice[] = await this.getPricesByAddress([address], currency);
+      tokenPrice = prices[0];
+    }
+    return { address, price: tokenPrice.price, timestamp: tokenPrice.timestamp };
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
```

Comment:
I'm a bit confused about how these methods are laid out. They are all public, and it isn't indicated if any are _not_ supposed to be called by a user.

- getPriceByAddress
  - Checks the cache, returns the cache value if valid.
  - If cache is empty, calls getPricesByAddress to get the price
  - Returns the requested price.
- getPricesByAddress
  - Requests the price from the cache and the source.
  - Updates the cache with the source value.
  - Returns the requested prices.

Is getPricesByAddress not expected to be called by an external user? It feels like it's being used as an internal utility function. If so, it should probably be private or protected. However, both seem useful for a user. I could imagine a different setup where getPricesByAddress does all the cache checking and logging and getPriceByAddress is just a convenience function to forward the call.

Reply:
The intention is for both to be callable by the user, but it's true that only the singular form (`getPriceByAddress`) checks for a pre-existing valid price. I have an open observation about there [here](https://github.com/across-protocol/sdk-v2/pull/98#discussion_r968995512).

As you've described, it does seem better to migrate the cache liveness check to the plural form (`getPricesByAddress`) and implement a loop over the requested addresses to see whether they all have "live" prices. I'd avoided doing this because I didn't like implementing the extra loop, but it feels awkward that the two methods behave differently in this respect.

Reply:
@mrice32 Per the current version of this PR, I think we should be OK on this point @ :+1:.